tests/net/mod.rs

```diff
@@ -994,20 +988,53 @@ where
     C: Contribution + Clone,
     N: NodeIdT,
 {
-    /// Verifies that all nodes' outputs agree, and returns the output.
+    /// Verifies that all nodes' outputs agree, and returns the output of node 1. Requirements
+    /// towards the output from node 0 are weaker compared to other nodes: it can be removed - hence
```

Comment:
Let's not hard-code into the test frameworks that node 0 was removed. I hope to add more complicated tests at some point, that add and remove more nodes.
Maybe we should instead pass in one node's index (or ID?) that is expected to have received all the batches. (A test can always make sure such a node exists, even if it's just an observer.)
We'd then compute the expected output of each node by filtering the complete output, and omitting all the batches where that node wasn't present, according to `SenderQueueableOutput::added_peers`?

Reply:
That's right! I forgot to pass the node ID here...

Reply:
> Maybe we should instead pass in one node's index (or ID?) that is expected to have received all the batches. (A test can always make sure such a node exists, even if it's just an observer.) We'd then compute the expected output of each node by filtering the complete output, and omitting all the batches where that node wasn't present,

I implemented this, with a small difference that I pass the whole node by reference instead of by ID or index in the vector of correct nodes. I interpreted "node's presence" as whether the full node has a contribution for the node in question in a given epoch.

>  according to SenderQueueableOutput::added_peers?

I'm still thinking how better to add this to `verify_batches`. Or maybe in a separate check?

Reply:
Yes, I think it can't be done in a simple `filter` call. You'd need to essentially mimick what `SenderQueue` is doing, for every node:
* The first batch is expected. (This will also have to change at some point, for nodes that are added later.)
* Whenever `SenderQueueableOutput::added_peers` returns a new set that doesn't contain the node, all _following_ batches are not expected in the output.
* Whenever it returns a new set that _does_ contain the node, all following batches are expected again.

Not sure whether there's an easier way to do this. :confused:

Reply:
My case for a separate check is this: if we check every batch against `SenderQueueableOutput::added_peers`, we would get early stage errors instead of divergence and an uninformative `MessageLimitExceeded`.

Waiting until all the batches are output is not helpful if the test diverges due to invalid batches.

Reply:
Right, that makes sense.
We should still test for both, though, and also assert that at the end, everyone got all the expected batches.